executor/executor_fuchsia.h

```diff
@@ -6,9 +6,107 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <zircon/process.h>
 #include <zircon/syscalls.h>
 
-#include "nocover.h"
+#define MAX_COVSZ (1ULL << 20)
+
+// In x86_64, sancov stores the return address - 1.
+// We add 1 so the stored value points to a valid pc.
+static const uint64_t kPcFixup = 1;
+
+static const char* kCovPcsFileName = "/boot/kernel/data/zircon.elf.1.sancov";
+
+typedef struct cover_ctx_t {
+	uint64_t base_covcount[MAX_COVSZ];
+	uint64_t curr_covcount[MAX_COVSZ];
+	uint64_t pc_table[MAX_COVSZ];
+	uint32_t real_coverage_truncated[MAX_COVSZ];
+	size_t total_pcs; // number of elements in pc_table and covcount tables, determined by kernel.
+	zx_handle_t covcount_vmo;
+} cover_ctx;
+
+static __thread cover_ctx cover;
+
+static void cover_open(cover_t* cov, bool extra)
+{
+}
+
+static void cover_enable(cover_t* cov, bool collect_comps, bool extra)
+{
+	zx_status_t status = zx_coverage_ctl(zx_thread_self(), 1, ZX_HANDLE_INVALID);
+	if (status != ZX_OK) {
+		fail("failed to enable coverage. err: %d\n", status);
+	}
+
+	status = zx_vmo_create(sizeof(cover.curr_covcount), 0, &cover.covcount_vmo);
+	if (status != ZX_OK) {
+		fail("failed to create covcount vmo. err: %d\n", status);
+	}
+}
+
+static size_t snapshot_sancov(uint64_t* dst, size_t elems, const char* filename)
+{
+	FILE* f = fopen(filename, "rb");
+	if (f == NULL) {
+		fail("could not open coverage file '%s'", filename);
+	}
+
+	size_t n = fread(dst, sizeof(uint64_t), elems, f);
+	if (n == elems) {
+		fail("pc table is too small. make it bigger.");
+	}
+
+	fclose(f);
+	return n;
+}
+
+static size_t snapshot_pctable(uint64_t* dst, size_t elems)
+{
+	return snapshot_sancov(dst, elems, kCovPcsFileName);
+}
+
+static void snapshot_covcount(uint64_t* dst, size_t elems)
+{
+	// TODO: Only read the right amount of coverage.
+	zx_status_t status = zx_coverage_ctl(zx_thread_self(), 2, cover.covcount_vmo);
+	if (status != ZX_OK) {
+		fail("failed to fetch coverage. err: %d\n", status);
+	}
+	status = zx_vmo_read(cover.covcount_vmo, dst, 0, elems * sizeof(uint64_t));
+	if (status != ZX_OK) {
+		fail("failed to copy coverage. err: %d\n", status);
+	}
+}
+
+static void cover_reset(cover_t* cov)
+{
+	snapshot_covcount(cover.base_covcount, MAX_COVSZ);
+}
+
+static void cover_collect(cover_t* cov)
```

Comment:
As far as I remember this can be called not on the target thread in some cases (blocked syscalls), so I think this needs to be using the cov argument in some way.